File: libs/go/types/api/responses/subscription_event.go
package responses

import (
	"time"

	"github.com/google/uuid"
)

// SubscriptionEventPriceInfo contains essential price information relevant to a subscription event
type SubscriptionEventPriceInfo struct {
	ID                  uuid.UUID `json:"id"`
	Type                string    `json:"type"`
	Currency            string    `json:"currency"`
	UnitAmountInPennies int64     `json:"unit_amount_in_pennies"`
	IntervalType        string    `json:"interval_type,omitempty"`
	TermLength          int32     `json:"term_length,omitempty"`
	CreatedAt           int64     `json:"created_at"`
	UpdatedAt           int64     `json:"updated_at"`
}

// SubscriptionEventResponse represents a detailed view of a subscription event
type SubscriptionEventResponse struct {
	ID                 uuid.UUID                  `json:"id"`
	SubscriptionID     uuid.UUID                  `json:"subscription_id"`
	EventType          string                     `json:"event_type"`
	TransactionHash    string                     `json:"transaction_hash,omitempty"`
	EventAmountInCents int32                      `json:"event_amount_in_cents"`
	EventOccurredAt    time.Time                  `json:"event_occurred_at"`
	ErrorMessage       string                     `json:"error_message,omitempty"`
	EventMetadata      []byte                     `json:"event_metadata,omitempty"`
	EventCreatedAt     time.Time                  `json:"event_created_at"`
	CustomerID         uuid.UUID                  `json:"customer_id,omitempty"`
	SubscriptionStatus string                     `json:"subscription_status,omitempty"`
	ProductID          uuid.UUID                  `json:"product_id,omitempty"`
	ProductName        string                     `json:"product_name,omitempty"`
	Customer           CustomerResponse           `json:"customer"`
	Product            ProductResponse            `json:"product"`
	PriceInfo          SubscriptionEventPriceInfo `json:"price_info"`
	ProductToken       ProductTokenResponse       `json:"product_token"`
	Network            NetworkResponse            `json:"network"`
}

type SubscriptionEventFullResponse struct {
	ID                 uuid.UUID                  `json:"id"`
	SubscriptionID     uuid.UUID                  `json:"subscription_id"`
	EventType          string                     `json:"event_type"`
	TransactionHash    string                     `json:"transaction_hash,omitempty"`
	EventAmountInCents int32                      `json:"event_amount_in_cents"`
	EventOccurredAt    time.Time                  `json:"event_occurred_at"`
	ErrorMessage       string                     `json:"error_message,omitempty"`
	EventMetadata      []byte                     `json:"event_metadata,omitempty"`
	EventCreatedAt     time.Time                  `json:"event_created_at"`
	CustomerID         uuid.UUID                  `json:"customer_id,omitempty"`
	SubscriptionStatus string                     `json:"subscription_status,omitempty"`
	ProductID          uuid.UUID                  `json:"product_id,omitempty"`
	ProductName        string                     `json:"product_name,omitempty"`
	Customer           CustomerResponse           `json:"customer"`
	Product            ProductResponse            `json:"product"`
	PriceInfo          SubscriptionEventPriceInfo `json:"price_info"`
	ProductToken       ProductTokenResponse       `json:"product_token"`
	Network            NetworkResponse            `json:"network"`
}

// ListSubscriptionEventsResult represents the result of listing subscription events
type ListSubscriptionEventsResult struct {
	Events []SubscriptionEventFullResponse `json:"events"`
	Total  int64                           `json:"total"`
}

File: apps/web-app/src/types/subscription-event.ts
import { CustomerResponse } from './customer';
import { NetworkResponse } from './network';
import { ProductTokenResponse } from './product';
import type { SubscriptionStatus } from './subscription';

/**
 * Subscription event/transaction types
 */
export type SubscriptionEventType = 'redeemed' | 'failed' | 'failed_redemption';

/**
 * Contains essential price information relevant to a subscription event.
 * Simplified compared to a full PriceResponse for embedding.
 */
export interface SubscriptionEventPriceInfo {
  id: string; // uuid.UUID
  type: string; // db.PriceType (e.g., recurring, one_off)
  currency: string; // db.Currency
  unit_amount_in_pennies: number; // int32
  interval_type?: string; // db.NullIntervalType (optional)
  interval_count?: number; // pgtype.Int4 (optional)
  term_length?: number; // pgtype.Int4 (optional)
}

/**
 * Represents a detailed view of a subscription event.
 * Includes denormalized information from related entities like subscription, product, and price.
 */
export interface SubscriptionEventFullResponse {
  id: string; // uuid.UUID
  subscription_id: string; // uuid.UUID
  event_type: SubscriptionEventType; // db.SubscriptionEventType
  transaction_hash?: string; // pgtype.Text (optional)
  event_amount_in_cents: number; // int32
  event_occurred_at: string; // pgtype.Timestamptz
  error_message?: string; // pgtype.Text (optional)
  event_metadata?: Record<string, unknown> | null; // json.RawMessage (optional)
  event_created_at: string; // pgtype.Timestamptz
  customer_id: string; // uuid.UUID
  subscription_status: SubscriptionStatus; // db.SubscriptionStatus
  product_id: string; // uuid.UUID
  product_name: string;
  price_info: SubscriptionEventPriceInfo;
  product_token: ProductTokenResponse;
  network: NetworkResponse;
  customer: CustomerResponse;
}
